In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df = pd.read_csv('/home/bishnu/project/EDA_transaction_data/data/raw/transactions.csv')

In [4]:
df

accountNumber  customerId  creditLimit  availableMoney  \
0           733493772   733493772         5000         5000.00   
1           733493772   733493772         5000         4888.67   
2           733493772   733493772         5000         4863.92   
3           733493772   733493772         5000         4676.52   
4           733493772   733493772         5000         4449.18   
...               ...         ...          ...             ...   
641909      186770399   186770399         7500         2574.02   
641910      186770399   186770399         7500         2568.65   
641911      186770399   186770399         7500         2344.95   
641912      186770399   186770399         7500         2206.53   
641913      186770399   186770399         7500         2190.22   

        transactionDateTime  transactionAmount              merchantName  \
0       2016-01-08T19:04:50             111.33                      Lyft   
1       2016-01-09T22:32:39              24.75                      Uber   
2       2016-01-11T13:36:55             187.40                      Lyft   
3       2016-01-11T22:47:46             227.34                      Lyft   
4       2016-01-16T01:41:11               0.00                      Lyft   
...                     ...                ...                       ...   
641909  2016-12-04T12:29:21               5.37              Apple iTunes   
641910  2016-12-09T04:20:35             223.70      Blue Mountain eCards   
641911  2016-12-16T07:58:23             138.42             Fresh Flowers   
641912  2016-12-19T02:30:35              16.31                   abc.com   
641913  2016-12-28T11:14:14              32.53  Next Day Online Services   

       acqCountry merchantCountryCode  posEntryMode  ...  accountOpenDate  \
0              US                  US           5.0  ...         3/8/2014   
1              US                  US           9.0  ...         3/8/2014   
2              US                  US           5.0  ...         3/8/2014   
3              US                  US           2.0  ...         3/8/2014   
4              US                  US           2.0  ...         3/8/2014   
...           ...                 ...           ...  ...              ...   
641909         US                  US           5.0  ...        4/11/2015   
641910         US                  US           9.0  ...        4/11/2015   
641911         US                  US           2.0  ...        4/11/2015   
641912         US                  US           9.0  ...        4/11/2015   
641913         US                  US           9.0  ...        4/11/2015   

       dateOfLastAddressChange cardCVV enteredCVV cardLast4Digits  \
0                     3/8/2014     492        492            9184   
1                     3/8/2014     492        492            9184   
2                     3/8/2014     492        492            9184   
3                     3/8/2014     492        492            9184   
4                     3/8/2014     492        492            9184   
...                        ...     ...        ...             ...   
641909                3/6/2016     127        127            5432   
641910                3/6/2016     127        127            5432   
641911                3/6/2016     127        127            5432   
641912                3/6/2016     127        127            5432   
641913                3/6/2016     127        127            5432   

             transactionType  isFraud  currentBalance cardPresent  \
0                   PURCHASE     True            0.00       False   
1                   PURCHASE    False          111.33       False   
2                   PURCHASE    False          136.08       False   
3                   PURCHASE     True          323.48       False   
4       ADDRESS_VERIFICATION    False          550.82       False   
...                      ...      ...             ...         ...   
641909              PURCHASE    False         4925.98       False

In [5]:
df.shape

(641914, 23)

In [6]:
df.sample(5)

accountNumber  customerId  creditLimit  availableMoney  \
60485       235721673   235721673         5000         2894.44   
225102      990764813   990764813        20000        11497.03   
4129        513859788   513859788         2500         1050.04   
254835      129149357   129149357         1000         1000.00   
607224      845992857   845992857          500          199.09   

        transactionDateTime  transactionAmount          merchantName  \
60485   2016-06-17T14:04:40             542.01            AMC #79863   
225102  2016-08-03T15:04:24             383.28  Krispy Kreme #685312   
4129    2016-05-13T13:18:59              37.25     Mobil Gas #851671   
254835  2016-06-01T17:30:01             189.48                  Lyft   
607224  2016-03-20T15:05:59              19.90             sears.com   

       acqCountry merchantCountryCode  posEntryMode  ...  accountOpenDate  \
60485          US                  US           9.0  ...        15/6/2013   
225102         US                  US           9.0  ...        10/1/2015   
4129           US                  US           9.0  ...       25/11/2015   
254835         US                  US           5.0  ...        21/7/2015   
607224         US                  US           5.0  ...        19/1/2015   

       dateOfLastAddressChange cardCVV enteredCVV cardLast4Digits  \
60485                27/5/2016     397        397            9984   
225102               12/6/2016     910        910            4008   
4129                25/11/2015     553        553            5655   
254835               21/7/2015     207        207             915   
607224               19/1/2015     196        196             233   

        transactionType  isFraud  currentBalance cardPresent  \
60485          PURCHASE    False         2105.56        True   
225102         PURCHASE    False         8502.97        True   
4129           PURCHASE    False         1449.96        True   
254835         PURCHASE    False            0.00       False   
607224         PURCHASE    False          300.91       False   

        expirationDateKeyInMatch  
60485                      False  
225102                     False  
4129                       False  
254835                     False  
607224                     False  

[5 rows x 23 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 641914 entries, 0 to 641913
Data columns (total 23 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   accountNumber             641914 non-null  int64  
 1   customerId                641914 non-null  int64  
 2   creditLimit               641914 non-null  int64  
 3   availableMoney            641914 non-null  float64
 4   transactionDateTime       641914 non-null  object 
 5   transactionAmount         641914 non-null  float64
 6   merchantName              641914 non-null  object 
 7   acqCountry                638001 non-null  object 
 8   merchantCountryCode       641290 non-null  object 
 9   posEntryMode              638569 non-null  float64
 10  posConditionCode          641627 non-null  float64
 11  merchantCategoryCode      641914 non-null  object 
 12  currentExpDate            641914 non-null  object 
 13  accountOpenDate           641914 non-null  o

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.isna().sum()

accountNumber                  0
customerId                     0
creditLimit                    0
availableMoney                 0
transactionDateTime            0
transactionAmount              0
merchantName                   0
acqCountry                  3913
merchantCountryCode          624
posEntryMode                3345
posConditionCode             287
merchantCategoryCode           0
currentExpDate                 0
accountOpenDate                0
dateOfLastAddressChange        0
cardCVV                        0
enteredCVV                     0
cardLast4Digits                0
transactionType              589
isFraud                        0
currentBalance                 0
cardPresent                    0
expirationDateKeyInMatch       0
dtype: int64

In [10]:
df[df["transactionType"].isna()]

accountNumber  customerId  creditLimit  availableMoney  \
715         814139734   814139734         7500         7500.00   
1025        885773130   885773130         7500         4421.60   
1064        799731758   799731758        50000        49979.93   
2372        738735439   738735439          500           33.59   
2784        738735439   738735439          500          424.21   
...               ...         ...          ...             ...   
637606      240456714   240456714        15000         7367.49   
638589      953570469   953570469        10000         6545.58   
640301      584174368   584174368         1000          490.47   
640764      584174368   584174368         1000           70.17   
641780      455115932   455115932        50000        36162.74   

        transactionDateTime  transactionAmount           merchantName  \
715     2016-04-03T09:08:47              28.03               ebay.com   
1025    2016-11-30T22:27:48             103.93                   Uber   
1064    2016-07-19T03:21:18             238.63             Best Fries   
2372    2016-03-05T20:45:01               4.11         Subway #197429   
2784    2016-08-28T07:16:35              37.79      McDonalds #927989   
...                     ...                ...                    ...   
637606  2016-09-06T13:58:02              64.37            westelm.com   
638589  2016-10-30T06:26:47             102.18            NYSC #67113   
640301  2016-03-30T10:26:14               8.77            Dallas Post   
640764  2016-11-11T06:42:16              42.81            Dallas Post   
641780  2016-08-20T08:20:20               2.12  Fresh Online Services   

       acqCountry merchantCountryCode  posEntryMode  ...  accountOpenDate  \
715            US                  US           9.0  ...        21/9/2011   
1025           US                  US           2.0  ...         9/4/2015   
1064           US                  US           9.0  ...       16/10/2015   
2372           US                  US           9.0  ...        15/2/2014   
2784           US                  US           9.0  ...        15/2/2014   
...           ...                 ...           ...  ...              ...   
637606         US                  US           9.0  ...        12/8/2014   
638589         US                  US           2.0  ...        7/12/2013   
640301         US                  US           5.0  ...        10/3/2015   
640764         US                  US           2.0  ...        10/3/2015   
641780         US                  US           9.0  ...         2/2/2011   

       dateOfLastAddressChange cardCVV enteredCVV cardLast4Digits  \
715                  21/9/2011     280        280            1744   
1025                  6/7/2016     612        612            9852   
1064                16/10/2015     159        159            7155   
2372                 15/2/2014     729        729            6074   
2784                 20/5/2016     729        729            6074   
...                        ...     ...        ...             ...   
637606               22/8/2016     439        439            9487   
638589               7/12/2013     350        350               7   
640301               10/3/2015     893        893            3586   
640764               30/8/2016     893        893            3586   
641780                2/2/2011     148        148            2521   

        transactionType  isFraud  currentBalance cardPresent  \
715                 NaN    False            0.00       False   
1025                NaN    False         3078.40       False   
1064                NaN    False           20.07        True   
2372                NaN    False          466.41       False   
2784                NaN    False           75.79        True   
...                 ...      ...             ...         ...   
637606              NaN    False         7632.51        True   
638589              NaN     True         3454.42       False   
640301  

In [11]:
df["customerId"].value_counts().reset_index()

customerId  count
0      318001076  10034
1      456044564   8382
2      812328116   5494
3      838085703   5129
4      239875038   4705
...          ...    ...
4995   800578637      1
4996   156291319      1
4997   845208594      1
4998   703783494      1
4999   631951213      1

[5000 rows x 2 columns]

- Columns having null values:
- acqCountry (3913)
- merchantCountryCode (624)
- posEntryMode (3345)
- posConditionCode (287)
- transactionType (589)


In [ ]:
def plot_bar_graph(data,x,y,w,h,tick=False):
  plt.figure(figsize=(w,h))
  ax = sns.barplot(data=data,x=x,y=y,order=data[x])
  for i in ax.containers:
    ax.bar_label(i,)
  plt.ylabel(y+(" (%)"))
  if tick:
    plt.xticks(rotation=45)
  plt.show()

In [12]:
df.columns

Index(['accountNumber', 'customerId', 'creditLimit', 'availableMoney',
       'transactionDateTime', 'transactionAmount', 'merchantName',
       'acqCountry', 'merchantCountryCode', 'posEntryMode', 'posConditionCode',
       'merchantCategoryCode', 'currentExpDate', 'accountOpenDate',
       'dateOfLastAddressChange', 'cardCVV', 'enteredCVV', 'cardLast4Digits',
       'transactionType', 'isFraud', 'currentBalance', 'cardPresent',
       'expirationDateKeyInMatch'],
      dtype='object')

In [13]:
df["acqCountry"].value_counts()

acqCountry
US     632303
MEX      2626
CAN      1870
PR       1202
Name: count, dtype: int64

In [14]:
df["merchantCountryCode"].value_counts()

merchantCountryCode
US     635577
MEX      2636
CAN      1874
PR       1203
Name: count, dtype: int64

In [15]:
new_df = df[~((df["acqCountry"].isna()) & (df["merchantCountryCode"].isna()))]

In [16]:
new_df[new_df["merchantCountryCode"].isna()]

accountNumber  customerId  creditLimit  availableMoney  \
427536      838085703   838085703         5000         1873.01   

        transactionDateTime  transactionAmount           merchantName  \
427536  2016-05-26T18:53:06             393.36  Regal Cinemas #276341   

       acqCountry merchantCountryCode  posEntryMode  ...  accountOpenDate  \
427536         US                 NaN          80.0  ...        24/9/2015   

       dateOfLastAddressChange cardCVV enteredCVV cardLast4Digits  \
427536               22/4/2016     625        625            4062   

        transactionType  isFraud  currentBalance cardPresent  \
427536         PURCHASE    False         3126.99        True   

        expirationDateKeyInMatch  
427536                     False  

[1 rows x 23 columns]

In [17]:
new_df["merchantCountryCode"].fillna("US", inplace=True)

/tmp/ipykernel_27058/3860493255.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  new_df["merchantCountryCode"].fillna("US", inplace=True)
/tmp/ipykernel_27058/3860493255.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["merchantCountryCode"].fillna("US", inplace=True)


In [18]:
new_df["acqCountry"].value_counts()

acqCountry
US     632303
MEX      2626
CAN      1870
PR       1202
Name: count, dtype: int64

In [19]:
new_df["merchantCountryCode"].value_counts()

merchantCountryCode
US     635578
MEX      2636
CAN      1874
PR       1203
Name: count, dtype: int64

In [20]:
acq_merchant_country = new_df[new_df["acqCountry"].isna()][["acqCountry","merchantCountryCode"]]

In [21]:
acq_merchant_country["merchantCountryCode"].value_counts()

merchantCountryCode
US     3268
MEX      11
CAN       9
PR        2
Name: count, dtype: int64

In [22]:
new_df.shape

(641291, 23)

In [23]:
def impute_acqCountry(row):
    if pd.isna(row["acqCountry"]):
        if row["merchantCountryCode"] == "US":
            return "US"
        elif row["merchantCountryCode"] == "MEX":
            return "MEX"
        elif row["merchantCountryCode"] == "CAN":
            return "CAN"
        else:
            return "PR"
    return row["acqCountry"]

In [24]:
new_df["acqCountry"] = new_df.apply(impute_acqCountry, axis=1)

/tmp/ipykernel_27058/3960673797.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["acqCountry"] = new_df.apply(impute_acqCountry, axis=1)


In [25]:
new_df.shape

(641291, 23)

In [26]:
def check(row):
  count = 0
  if row["acqCountry"] != row["merchantCountryCode"]:
    count = 1
  return count

In [27]:
new_df['cross_border_count'] = new_df.apply(check, axis=1)

/tmp/ipykernel_27058/1105594127.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['cross_border_count'] = new_df.apply(check, axis=1)


In [28]:
new_df['cross_border_count'].value_counts()

cross_border_count
0    641230
1        61
Name: count, dtype: int64

In [29]:
new_df["transactionType"].value_counts()

transactionType
PURCHASE                608087
ADDRESS_VERIFICATION     16466
REVERSAL                 16151
Name: count, dtype: int64

In [30]:
# Impute transactionType by highest frequent value
new_df["transactionType"].fillna("PURCHASE", inplace=True)

/tmp/ipykernel_27058/3461299649.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  new_df["transactionType"].fillna("PURCHASE", inplace=True)
/tmp/ipykernel_27058/3461299649.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["transactionType"].fillna("PURCHASE", inplace=True)


In [31]:
# Rows containing missing values for "posEntryMode" and "posConditionCode" are removed
new_df.dropna(axis=0, inplace=True)

/tmp/ipykernel_27058/2269763502.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df.dropna(axis=0, inplace=True)


In [32]:
new_df['cross_border_count'] = new_df.apply(check, axis=1)

/tmp/ipykernel_27058/1105594127.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['cross_border_count'] = new_df.apply(check, axis=1)


In [33]:
new_df['cross_border_count'].value_counts()

cross_border_count
0    637606
1        60
Name: count, dtype: int64

In [34]:
new_df.isna().sum()


accountNumber               0
customerId                  0
creditLimit                 0
availableMoney              0
transactionDateTime         0
transactionAmount           0
merchantName                0
acqCountry                  0
merchantCountryCode         0
posEntryMode                0
posConditionCode            0
merchantCategoryCode        0
currentExpDate              0
accountOpenDate             0
dateOfLastAddressChange     0
cardCVV                     0
enteredCVV                  0
cardLast4Digits             0
transactionType             0
isFraud                     0
currentBalance              0
cardPresent                 0
expirationDateKeyInMatch    0
cross_border_count          0
dtype: int64

In [35]:
new_df.shape

(637666, 24)

In [37]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 637666 entries, 0 to 641913
Data columns (total 24 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   accountNumber             637666 non-null  int64  
 1   customerId                637666 non-null  int64  
 2   creditLimit               637666 non-null  int64  
 3   availableMoney            637666 non-null  float64
 4   transactionDateTime       637666 non-null  object 
 5   transactionAmount         637666 non-null  float64
 6   merchantName              637666 non-null  object 
 7   acqCountry                637666 non-null  object 
 8   merchantCountryCode       637666 non-null  object 
 9   posEntryMode              637666 non-null  float64
 10  posConditionCode          637666 non-null  float64
 11  merchantCategoryCode      637666 non-null  object 
 12  currentExpDate            637666 non-null  object 
 13  accountOpenDate           637666 non-null  object

In [38]:
# changing datetime columun to datetime format
new_df['transactionDateTime'] = pd.to_datetime(new_df['transactionDateTime'])
new_df['transactionDateTime'].dtype

/tmp/ipykernel_27058/3696934381.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['transactionDateTime'] = pd.to_datetime(new_df['transactionDateTime'])


dtype('<M8[ns]')

In [47]:
new_df.columns

Index(['accountNumber', 'customerId', 'creditLimit', 'availableMoney',
       'transactionDateTime', 'transactionAmount', 'merchantName',
       'acqCountry', 'merchantCountryCode', 'posEntryMode', 'posConditionCode',
       'merchantCategoryCode', 'currentExpDate', 'accountOpenDate',
       'dateOfLastAddressChange', 'cardCVV', 'enteredCVV', 'cardLast4Digits',
       'transactionType', 'isFraud', 'currentBalance', 'cardPresent',
       'expirationDateKeyInMatch', 'cross_border_count'],
      dtype='object')

In [49]:
df.shape

(641914, 23)

In [48]:
new_df.shape

(637666, 24)

In [51]:
# Extracting year, month, and day from the transactionDateTime column
new_df['txn_hour'] = new_df['transactionDateTime'].dt.hour
new_df['txn_month'] = new_df['transactionDateTime'].dt.month
new_df['txn_day'] = new_df['transactionDateTime'].dt.day

/tmp/ipykernel_27058/262847115.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['txn_hour'] = new_df['transactionDateTime'].dt.hour
/tmp/ipykernel_27058/262847115.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['txn_month'] = new_df['transactionDateTime'].dt.month
/tmp/ipykernel_27058/262847115.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: ht

In [40]:
new_df["currentExpDate"] = pd.to_datetime(new_df["currentExpDate"], dayfirst=True)

/tmp/ipykernel_27058/731969811.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["currentExpDate"] = pd.to_datetime(new_df["currentExpDate"], dayfirst=True)


In [42]:
new_df["currentExpDate"].dtype

dtype('<M8[ns]')

In [43]:
new_df["accountOpenDate"] = pd.to_datetime(new_df["accountOpenDate"], dayfirst=True)

/tmp/ipykernel_27058/1417457055.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["accountOpenDate"] = pd.to_datetime(new_df["accountOpenDate"], dayfirst=True)


In [50]:
new_df["accountOpenDate"].dtype

dtype('<M8[ns]')

In [45]:
new_df.shape

(637666, 27)

In [55]:
new_df.to_csv('/home/bishnu/project/EDA_transaction_data/data/processed/transactions_cleaned.csv', index=False)

* Duplicates: 0 rows
* Missing values:
    • acqCountry (3913)
    • merchantCountryCode (624)
    • posEntryMode (3345)
    • posConditionCode (287)
    • transactionType (589)
* Handling Missing Values
    - Since ”acqCountry” and ”merchantCountrycode” contains same values for each row
but there are around 623 rows which have missing values for both columns. Hence
these rows are removed.
    - Remaining value for ”merchantCountryCode” is imputed with the help of ”acqCountry” and vice-versa
    - Missing ”transactionType” values are imputed by most frequent type i.e ”PURCHASE”
    - Rows containing missing values for ”posEntryMode” and ”posConditionCode” are
removed.
* DateTime columns are converted into pandas date time format.
* New features are added as day, hour and month of transactionDateTime.
* cross_border_count feature is added, which represents different merchantCountryCode and acqCountry for a transaction.
* Final dataset is converted into shape of (637666, 27).

In [54]:
new_df.shape

(637666, 27)